hamilton/function_modifiers/macros.py

```diff
@@ -1055,3 +1055,127 @@ def chain_transforms(
             )
             first_arg = raw_node.name
     return nodes, first_arg
+
+
+class NotSameModuleError(Exception):
+    def __init__(self, fn, target_fn):
+        super().__init__(
+            f"The functions have to be in the same module... "
+            f"The target function {target_fn.__name__} is in module {target_fn.__module__} and "
+            f"the mutator function {fn.__name__} is in module {fn.__module__}./n"
+            "Use power user setting to disable this restriction."
+        )
+
+
+class mutate:
+    """Running a transformation on the outputs of a series of functions.
+
+    This is closely related to `pipe_output` as it effectively creates the option to construct `pipe_output`
+    remotely. We choose which target functions we wish to mutate by the transformation we are decorating.
+    For now, the target functions that will be mutated have to be in the same module (come speak to us
+    if you need this capability over multiple modules).
+
+    If we wish to apply `_transform1` to the output of A and B and `_transform2` only to the output
+    of node B, we can do this like
+
+    .. code-block:: python
+        :name: Simple @mutate example
+
+        def A(...):
+            return ...
+
+        def B(...):
+            return ...
+
+        @mutate(A,B)
+        def _transform1(...):
+            return ...
+
+        @mutate(B)
+        def _transform2(...):
+            return ...
+
+    we obtain the new pipe-like subDAGs **A_raw --> _transform1 --> A** and **B_raw --> _transform1 --> _transform2 --> B**,
+    where the behavior is the same as `pipe_output`.
+
+    While it is generally reasonable to contain these constructs within a node's function,
+    you should consider `mutate` in the following scenarios:
+
+    1. Loading data and applying pre-cleaning step.
+    2. Feature engineering via joining, filtering, sorting, etc.
+    3. Experimenting with different transformations across nodes by selectively turning transformations on / off.
+
+    We assume the first argument of the decorated function to be the output of the function we are targeting.
+    For transformations with multiple arguments you can use key word arguments couples with `step` or `value`
+    the same as with other `pipe`-family decorators
+
+    .. code-block:: python
+        :name: @mutate example with multiple arguments
+
+        @mutate(A,B, arg2=step('upstream_node'),arg3=value(some_literal),...)
+        def _transform1(output_from_target:correct_type, arg2:arg2_type, arg3:arg3_type,...):
+            return ...
+
+    """
+
+    def __init__(
+        self,
+        *target_functions: Callable,
```

Comment:
I think we might need to support a string -- since it could come from extract_fields for instance.

Reply:
That is an interesting edge case and also one for `pipe_output` that we haven't addressed;  right now if we try

```python
def _pre_step(something:int)->int:
        return something + 10

def _post_step(something:int)->int:
        return something + 100

def a()->int:
        return 10

@pipe_output(
        step(_pre_step)
)
@extract_fields(
        {"field_1":int, "field_2":int}
)
@pipe_input(
        step(_pre_step)
)
def foo(a:int)->Dict[str,int]:
    return {"field_1":1, "field_2":2}
```
we run into an `InvalidDecoratorException: Expected a single node to transform, but got 2.` when trying to use `post_pipe`. Since `mutate` is an extension of `post_pipe` we get the same issue.

I think this should be solved upstream in `post_pipe` and added as as allowed option, what do you think?

Reply:
This also opens up the discussion from my notes, using with other decorators we need to decide on an individual basis when `mutate` is meant to be applied. For example `check_output` should be applied to the result of `mutate`, but here `mutate` should be applied to the result of `exteact_fields`. Sometimes you may even want to `mutate` the function first and then extract field?

Reply:
This is difficult to get right and I'd prefer not to support it for that reason. @jernejfrank your original implementation would have made this easier (it worked on strings), but we're really just reusing the pieces we have to not add additional concerns inside the function graph (as we have not really thought about he implication).

Specifically:
1. Decorators operate on a per-function basis
2. This adds the function-level decorators to functions
3. We don't know the node until *after* we process the. functions

While we could change that, it gets really tricky, and I think it'll spaghettify the code.

To unblock this I'd say we add a target for `@mutate` -- simliar to how we do others:

```python
@mutate(other_function, target_=...)
def ...
```

Reply:
Hmm just to make sure we're on the same wave length I'm thinking about this:

```python
@extract_columns("foo", "bar")
def data_set(...) -> pd.DataFrame:
     ...

# but I want to use mutate on `foo` and ` bar` -- how do I do that?
@mutate("foo", "bar")
def _normalize(col: pd.Series) -> pd.Series:
  ...

@extract_fields({"train": pd.DataFrame, "test": pd.DataFrame})
def training_data( ...) -> dict:
   ...

# but I want to apply the same transforms to each data set
@mutate("train", "test")
def _transforms(df: pd.DataFrame) -> pd.DataFrame:
   ...

```

Reply:
Otherwise, I don't think the decorator layering is an issue we should be that concerned about.

Reply:
@skrawcz yes, that's what I'm talking about. Specifically that's hard. We need to resolve before applying mutate, as we actually change the name of the nodes. That requires a second pass, which means we need to hardcode mutations. Possible (@jernejfrank had this), but it hacks up function graph, and I don't think it's worth it. 

Rather if we have a function pointer and include the string as a target, then it gets easier.

Reply:
@elijahbenizzy `target` has other meaning with other decorators, so I wouldn't use that here.

One attempt at your point:
```python
@extract_columns("foo", "bar")
def data_set(...) -> pd.DataFrame:
     ...

# but I want to use mutate on `foo` and ` bar` -- how do I do that?
@mutate((dataset, "foo"), (dataset, "bar"))
def _normalize(col: pd.Series) -> pd.Series:
  ...

@extract_fields({"train": pd.DataFrame, "test": pd.DataFrame})
def training_data( ...) -> dict:
   ...

# but I want to apply the same transforms to each data set
@mutate((training_data, "train"), (training_data, "test"))
def _transforms(df: pd.DataFrame) -> pd.DataFrame:
   ...
```
but function pointers break down with `@config.when` which is a common decorator:

```python
@extract_columns("foo", "bar")
@config.when(x="y")
def data_set__y(...) -> pd.DataFrame:
     ...

@extract_columns("foo", "bar")
@config.when(x="z")
def data_set__z(...) -> pd.DataFrame:
     ...

# but I want to use mutate on `foo` and ` bar` -- how do I do that?
@mutate((???, "foo"), (???, "bar"))
def _normalize(col: pd.Series) -> pd.Series:
  ...
```
Otherwise Hamilton should process the nodes in order they were encountered. So I don't think it needs to be two pass if that assumption is true.

Edge cases to consider:
 - what happens if node it depends on doesn't exist? e.g. via `@config.when` or user error, etc.

Reply:
I like @skrawcz's suggestion: `@mutate((funtion_pointer, node_string),...)`.

It allows us to get out of the spaghetti mess with pure strings. I couldn't find a reasonable way of doing it without either adding another `mutate_functions` collector at the `graph.py` level or letting the DAG be generated as is and then adding a second transversal through nodes searching for the specific string names. The main problem is that without the function pointer, we lose any locality where the node will be generated, so it is impossible to add them to the current node resolution without the above ways that get around that problem. 

API-wise we can have `@mutate(Callable | Tuple(Callable, str))` and if it is only a function pointer it gets added as a classical `pipe_output`, if it is a Tuple it delays construction until we reach the node level that can be added to it (now we get this info, because the function_pointer has it stored!)

Re: config -- I think that's fine. We just say this is the API

```python
# but I want to use mutate on `foo` and ` bar` -- how do I do that?
@mutate(
    (data_set__y, "foo"),
    (data_set__y, "bar"),
    (data_set__z, "foo"),
    (data_set__z, "bar")
    )
def _normalize(col: pd.Series) -> pd.Series:
    return ...
```
or to make it less verbose:

@mutate(
    (data_set__y, ["foo","bar"], some_dataset_specif_variable=value(10) ),
    (data_set__z, ["foo","bar"], some_dataset_specif_variable=source('upstream_node'))
    )
def _normalize(col: pd.Series, some_dataset_specif_variable:int) -> pd.Series:
    return col*some_dataset_specif_variable
```

and the fully flexible option:
```python
@mutate(
    (data_set__y, "foo", some_dataset_specif_variable=value(10)),
    (data_set__y, "bar", some_dataset_specif_variable=value(20)),
    (data_set__z, "foo", some_dataset_specif_variable=source('upstream_node1')),
    (data_set__z, "bar", some_dataset_specif_variable=source('upstream_node2'))
    )
def _normalize(col: pd.Series, some_dataset_specif_variable:int) -> pd.Series:
    return ...
```

Practically, I think this would then add a tag on the decorator level and we just need to be sure that it is the last `NodeTransformer` decorator during the node resolution. Then the above use-cases can be handles easily since we get nodes passed into and have all the info we need / the config stuff has already been resolved.

What do you think?

Reply:
> @elijahbenizzy `target` has other meaning with other decorators, so I wouldn't use that here.
> 
> One attempt at your point:
> 
> ```python
> @extract_columns("foo", "bar")
> def data_set(...) -> pd.DataFrame:
>      ...
> 
> # but I want to use mutate on `foo` and ` bar` -- how do I do that?
> @mutate((dataset, "foo"), (dataset, "bar"))
> def _normalize(col: pd.Series) -> pd.Series:
>   ...
> 
> @extract_fields({"train": pd.DataFrame, "test": pd.DataFrame})
> def training_data( ...) -> dict:
>    ...
> 
> # but I want to apply the same transforms to each data set
> @mutate((training_data, "train"), (training_data, "test"))
> def _transforms(df: pd.DataFrame) -> pd.DataFrame:
>    ...
> ```
> 
> but function pointers break down with `@config.when` which is a common decorator:
> 
> ```python
> @extract_columns("foo", "bar")
> @config.when(x="y")
> def data_set__y(...) -> pd.DataFrame:
>      ...
> 
> @extract_columns("foo", "bar")
> @config.when(x="z")
> def data_set__z(...) -> pd.DataFrame:
>      ...
> 
> # but I want to use mutate on `foo` and ` bar` -- how do I do that?
> @mutate((???, "foo"), (???, "bar"))
> def _normalize(col: pd.Series) -> pd.Series:
>   ...
> ```
> 
> Otherwise Hamilton should process the nodes in order they were encountered. So I don't think it needs to be two pass if that assumption is true. This feature is only 50% useful to Hamilton if it can only depend on function pointers.
> 
> Edge cases to consider:
> 
> * what happens if node it depends on doesn't exist? e.g. via `@config.when` or user error, etc.

@skrawcz @jernejfrank `target` is explicitly meant for this. Semantically it's identical. If we append a chain, `target` provides the node on which the chain is appended/nodes are modified. 

Furthermore, it's partially wired in. `@post_pipe` is a `SingleNodeNodeTransformer`. This, by default, means that it will apply to all the "sink" nodes (`target=None`). We limit it to just a single sink node, so we *don't* have to have target.

Instead, we should be able to alter these decorators to be a `NodeTransformer` (superclass of `SingleNodeNodeTransformer`) and thus take in target. And wiring through the target is then the exact same parameter. See [this](https://github.com/DAGWorks-Inc/hamilton/blob/acd2fc75692cfbe4c98eede3f5e89988c3950f1a/hamilton/function_modifiers/base.py#L413).

So IMO target is more consistent, and pretty much the same as tuples above. Should be straightforward to get this to work, *if* we want to support this.

Then there's the question of what to do if a target isn't applied -- my guess (and we'd have to test) is that if we make `pipe_input` and `pipe_output` a `NodeTransformer`, and use the default target (of `None` -- implies all "sink" nodes get decorated"), then we'll actually be able to just apply the same pipe operation to every column. Cool!

E.G.

```python
@extract_fields({"train": pd.DataFrame, "test": pd.DataFrame})
def training_data( ...) -> dict:
   ...

# but I want to apply the same transforms to each data set
@mutate(training_data) # will, by default, apply to each field
def _transforms(df: pd.DataFrame) -> pd.DataFrame:
   ...
```

Regarding `config`, I think we just need to specify both functions. We *may* be able to use a string for a function and look for anything in the same module to see if it matches (E.G. without the double underscore), but we should save this edge case for a bit later.

Otherwise, re: second pass, it's subtle as to why it has to be done. It's not because it carries references to other nodes, it's because running this decorator requires *changing* the other node's names. E.G. you have a node `foo`. Then you `pipe_output`  `foo` so you have `foo_raw` -> `foo`. Changing the name of `foo` to `foo_raw` is something that has to be done outside of the current decorator framework.

Conceptually, this is equivalent to allowing decorators to modify any part of the *entire* DAG, rather than individual subdags produced by functions, which makes it tougher to reason about, and requires n passes (one for each decorator). I'd prefer to avoid this for now, and I actually quite like the function-pointer approach.

Reply:
Another thought -- `target` is actually overused in Hamilton. We have:
1. `target` being the parameter we apply to (see data loaders)
2. `target` being the output node we use -- see data savers, data validators, check_output, tag, etc...)

(2) is the more used case, and corresponds internally, but we don't have to name it `target` if it's confusing. Just that we should use the internal concept of `target` to implement, and that'll make it easier.

An idea, re: config. I think this should work:

```python
@extract_columns("foo", "bar")
@config.when(x="y")
def data_set__y(...) -> pd.DataFrame:
     ...

@extract_columns("foo", "bar")
@config.when(x="z")
def data_set__z(...) -> pd.DataFrame:
     ...

# but I want to use mutate on `foo` and ` bar` -- how do I do that?
@mutate('data_set', "foo", "bar") # however you want to specify foo/bar, we have a few different patterns
def _normalize(col: pd.Series) -> pd.Series:
  ...
```

Specifically, if we have a string, it can refer to:
1. The name of the function in the same module as the current one
2. The name of the function without `__` (I think we might even track it...)

Reply:
> @skrawcz @jernejfrank `target` is explicitly meant for this. Semantically it's identical. If we append a chain, `target` provides the node on which the chain is appended/nodes are modified.
> 
> Furthermore, it's partially wired in. `@post_pipe` is a `SingleNodeNodeTransformer`. This, by default, means that it will apply to all the "sink" nodes (`target=None`). We limit it to just a single sink node, so we _don't_ have to have target.
> 
> Instead, we should be able to alter these decorators to be a `NodeTransformer` (superclass of `SingleNodeNodeTransformer`) and thus take in target. And wiring through the target is then the exact same parameter. See [this](https://github.com/DAGWorks-Inc/hamilton/blob/acd2fc75692cfbe4c98eede3f5e89988c3950f1a/hamilton/function_modifiers/base.py#L413).
> 
> So IMO target is more consistent, and pretty much the same as tuples above. Should be straightforward to get this to work, _if_ we want to support this.

Awesome, thanks @elijahbenizzy for drawing my attention to the superclass. That is a very clean way to resolve this dilemma and does exactly what we want to achieve with tuples. I'll go ahead and implement it quickly to have a concrete case that we can further discuss if something feels off and we initially missed.

> 
> Then there's the question of what to do if a target isn't applied -- my guess (and we'd have to test) is that if we make `pipe_input` and `pipe_output` a `NodeTransformer`, and use the default target (of `None` -- implies all "sink" nodes get decorated"), then we'll actually be able to just apply the same pipe operation to every column. Cool!
> 
> E.G.
> 
> ```python
> @extract_fields({"train": pd.DataFrame, "test": pd.DataFrame})
> def training_data( ...) -> dict:
>    ...
> 
> # but I want to apply the same transforms to each data set
> @mutate(training_data) # will, by default, apply to each field
> def _transforms(df: pd.DataFrame) -> pd.DataFrame:
>    ...
> ```

pipe_input I don't think we need to touch (it's a NodeInjector now) and I am not sure how we would use it with mutate, can you expand where it comes into play @elijahbenizzy ? 

But I did change pipe_output to the superclass NodeTransformer and happy to report that your above hunch is correct. In case no target is specified, at least for extract_field, it gets applied to both "train" and "test" (some minor caveats around adopting the name from the node instead of the function in case namespace is "...", but that should be ok).

> Regarding `config`, I think we just need to specify both functions. We _may_ be able to use a string for a function and look for anything in the same module to see if it matches (E.G. without the double underscore), but we should save this edge case for a bit later.

I will start off simple that you need to input both functions and if everything works/we are happy with the API, this can be changed at the end if that's alright with you?

Reply:
Great! You are correct, `@pipe_input` does not apply to this, although it *does* apply to the other definition of `target` -- adding a parameter `target` would make sense here IMO: https://github.com/DAGWorks-Inc/hamilton/blob/acd2fc75692cfbe4c98eede3f5e89988c3950f1a/hamilton/function_modifiers/macros.py#L750. 

But that's a separate issue, so I created this: https://github.com/DAGWorks-Inc/hamilton/issues/1161.

LMK what you need/when you want a review!